File: test/source/sheet/xstyleloader.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <test/sheet/xstyleloader.hxx>

#include <com/sun/star/beans/XPropertySet.hpp>
#include <com/sun/star/sheet/XSpreadsheetDocument.hpp>

#include <com/sun/star/container/XNameContainer.hpp>

#include <com/sun/star/style/XStyleLoader.hpp>
#include <com/sun/star/style/XStyleLoader2.hpp>

#include <com/sun/star/style/XStyle.hpp>

#include <rtl/ustring.hxx>
#include "cppunit/extensions/HelperMacros.h"

using namespace css;
using namespace css::uno;

namespace apitest {


void XStyleLoader::testLoadStylesFromURL()
{

  uno::Reference< sheet::XSpreadsheetDocument > xTargetDoc(init(), UNO_QUERY_THROW);

  OUString aFileURL = getTestURL();

  uno::Reference< style::XStyleFamiliesSupplier > xFamilySupplier (xTargetDoc, UNO_QUERY_THROW);
  uno::Reference< style::XStyleLoader > xTargetStyleLoader (xFamilySupplier->getStyleFamilies(), UNO_QUERY_THROW);

  uno::Sequence< beans::PropertyValue > aOptions = xTargetStyleLoader->getStyleLoaderOptions();

  xTargetStyleLoader->loadStylesFromURL(aFileURL, aOptions);

  checkStyleProperties(xFamilySupplier);

}

void XStyleLoader::testLoadStylesFromDocument()
{

  uno::Reference< sheet::XSpreadsheetDocument > xTargetDoc(init(), UNO_QUERY_THROW);

  uno::Reference< lang::XComponent > xSourceDoc (getSourceComponent(), UNO_QUERY_THROW);

  uno::Reference< style::XStyleFamiliesSupplier > xFamilySupplier (xTargetDoc, UNO_QUERY_THROW);
  uno::Reference< style::XStyleLoader2 > xTargetStyleLoader (xFamilySupplier->getStyleFamilies(), UNO_QUERY_THROW);

  uno::Sequence< beans::PropertyValue > aOptions = xTargetStyleLoader->getStyleLoaderOptions();

  xTargetStyleLoader->loadStylesFromDocument(xSourceDoc, aOptions);

  checkStyleProperties(xFamilySupplier);

}

void XStyleLoader::checkStyleProperties( uno::Reference< style::XStyleFamiliesSupplier > xFamilySupplier)
{
    // check if targetDocument has myStyle
  uno::Reference< container::XNameAccess > xFamilies(xFamilySupplier->getStyleFamilies(), UNO_QUERY_THROW);
  uno::Reference< container::XNameContainer > xCellStyles(xFamilies->getByName("CellStyles"), UNO_QUERY_THROW);

  CPPUNIT_ASSERT_MESSAGE("Style not imported", xCellStyles->hasByName("myStyle"));

  // test the backgroundcolor is correctly imported
  uno::Reference< style::XStyle > xMyStyle (xCellStyles->getByName("myStyle"), UNO_QUERY_THROW);
  uno::Reference< beans::XPropertySet > xPropSet (xMyStyle, UNO_QUERY_THROW);

  OUString aCellStyleName("CellBackColor");
  uno::Any aBackColor = xPropSet->getPropertyValue(aCellStyleName);
  uno::Any expectedBackColor(sal_Int32(16724787));

  CPPUNIT_ASSERT_EQUAL_MESSAGE("Wrong CellBackColor" , expectedBackColor, aBackColor);

  // test default pageStyle

  uno::Reference< container::XNameContainer > xPageStyles(xFamilies->getByName("PageStyles"), UNO_QUERY_THROW);
  uno::Reference<beans::XPropertySet> xPagePropSet(xPageStyles->getByName("Default"), UNO_QUERY_THROW);

  uno::Any aPageBackColor = xPagePropSet->getPropertyValue("BackColor");
  uno::Any expectedPageBackColor(sal_Int32(13434879));

  CPPUNIT_ASSERT_EQUAL_MESSAGE("Wrong page style BackColor" , expectedPageBackColor, aPageBackColor);
}

}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: sw/qa/python/xstyleloader.py
#! /usr/bin/env python
# -*- tab-width: 4; indent-tabs-mode: nil; py-indent-offset: 4 -*-
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import pathlib
import unittest

from org.libreoffice.unotest import UnoInProcess, makeCopyFromTDOC
from com.sun.star.beans import PropertyValue


class TestXStyleLoader(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls._uno = UnoInProcess()
        cls._uno.setUp()

    @classmethod
    def tearDownClass(cls):
        cls._uno.tearDown()

    def test_loadStyleFromStream(self):
        xDoc = self.__class__._uno.openEmptyWriterDoc()
        self.assertIsNotNone(xDoc)

        xServiceManager = self.__class__._uno.xContext.ServiceManager
        simpleFileAccess = xServiceManager.createInstance(
            "com.sun.star.ucb.SimpleFileAccess")
        xInputStream = simpleFileAccess.openFileRead(
            pathlib.Path(makeCopyFromTDOC("xstyleloader.odt")).as_uri())

        p1 = PropertyValue(Name="InputStream", Value=xInputStream)
        p2 = PropertyValue(Name="LoadTextStyles", Value=True)

        styles = xDoc.getStyleFamilies()
        styles.loadStylesFromURL("private:stream", [p1, p2])
        textStyles = styles.getByName("ParagraphStyles")
        self.assertTrue(textStyles.hasByName("Test_Template"))

        xDoc.close(True)


if __name__ == '__main__':
    unittest.main()

# vim: set shiftwidth=4 softtabstop=4 expandtab:
